mountstats.go

```diff
@@ -0,0 +1,552 @@
+	BytesSent int
+	// Number of bytes received for this operation, including RPC headers and payload.
+	BytesReceived int
+	// Duration all requests spent queued for transmission before they were sent.
+	CumulativeQueueTime time.Duration
+	// Duration it took to get a reply back after the request was transmitted.
+	CumulativeTotalResponseTime time.Duration
+	// Duration from when a request was enqueued to when it was completely handled.
+	CumulativeTotalRequestTime time.Duration
+}
+
+// A NFSTransportStats contains statistics for the NFS mount RPC requests and
+// responses.
+type NFSTransportStats struct {
+	// The local port used for the NFS mount.
+	Port int
+	// Number of times the client has had to establish a connection from scratch
+	// to the NFS server.
+	Bind int
+	// Number of times the client has made a TCP connection to the NFS server.
+	Connect int
+	// Duration (in jiffies, a kernel internal unit of time) the NFS mount has
+	// spent waiting for connections to the server to be established.
+	ConnectIdleTime int
+	// Duration since the NFS mount last saw any RPC traffic.
+	IdleTime time.Duration
+	// Number of RPC requests for this mount sent to the NFS server.
+	Sends int
+	// Number of RPC responses for this mount received from the NFS server.
+	Receives int
+	// Number of times the NFS server sent a response with a transaction ID
+	// unknown to this client.
+	BadTransactionIDs int
+	// A running counter, incremented on each request as the current difference
+	// ebetween sends and receives.
+	CumulativeActiveRequests int
+	// A running counter, incremented on each request by the current backlog
+	// queue size.
+	CumulativeBacklog int
+
+	// Stats below only available with stat version 1.1.
+
+	// Maximum number of simultaneously active RPC requests ever used.
+	MaximumRPCSlotsUsed int
+	// A running counter, incremented on each request as the current size of the
+	// sending queue.
+	CumulativeSendingQueue int
+	// A running counter, incremented on each request as the current size of the
+	// pending queue.
+	CumulativePendingQueue int
+}
+
+// parseMountStats parses a /proc/[pid]/mountstats file and returns a slice
+// of Mount structures containing detailed information about each mount.
+// If available, statistics for each mount are parsed as well.
+func parseMountStats(r io.Reader) ([]*Mount, error) {
+	const (
+		device            = "device"
+		statVersionPrefix = "statvers="
+
+		nfs3Type = "nfs"
+		nfs4Type = "nfs4"
+	)
+
+	var mounts []*Mount
+
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+		// Only look for device entries in this function
+		ss := strings.Fields(string(s.Bytes()))
+		if len(ss) == 0 || ss[0] != device {
+			continue
+		}
+
+		m, err := parseMount(ss)
+		if err != nil {
+			return nil, err
+		}
+
+		// Does this mount also possess statistics information?
+		if len(ss) > deviceEntryLen {
+			// Only NFSv3 and v4 are supported for parsing statistics
```

Comment:
The manpage mentions:

```
              (4)  Optional  statistics  and configuration information.  Currently (as at Linux 2.6.26), only NFS filesystems export information
                   via this field.
```

So as far as I can tell, no other filesystem type exposes statistics here.